docs/dev_101_services.md

```diff
@@ -197,3 +197,62 @@ If you need more control over the service call, you can also pass an async funct
 async def custom_set_sleep_timer(entity, service_call):
     await entity.set_sleep_timer(service_call.data['sleep_time'])
 ```
+
+## Response Data
+
+Services may optionally respond to a service call with data for powering more advanced automations. The use of response data for services is meant for
+cases that do not fit the Home Assistant state. For example, a response stream of repeated objects, or API would be better queried on demand or filtered such as a search.
+
+Example code:
+
+```python
+import datetime
+import voluptuous as vol
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, ServiceCall, ServiceResponse
+from homeassistant.helpers import config_validation as cv, entity_platform, service
+
+
+SEARCH_ITEMS_SERVICE_NAME = "search_items"
+SEARCH_ITEMS_SCHEMA = vol.Schema({
+    vol.Required("start"): datetime.datetime,
+    vol.Required("end"): datetime.datetime,
+})
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the platform."""
+
+    async def search_items(call: ServiceCall) -> ServiceResponse:
+        """Search in the date range and return the matching items."""
+        if not call.return_response:
+            raise ValueError("Request did not ask for response data")
+        items = await my_client.search(call.data["start"], call.data["end"])
+        return {
+            "items": [
+                {
+                    "summary": item["summary"],
+                    "description": item["description"],
+                } for item in items
+            ],
+        }
+
+      hass.services.async_register(
+          DOMAIN,
+          SEARCH_ITEMS_SERVICE_NAME,
+          search_items,
+          schema=SEARCH_ITEMS_SCHEMA,
+          supports_response=True,
+      )
+```
+
+There are some additional implementation standards:
+
+- You must set `supports_response` when registering a service so that the core
+knows which services support response data.
+- All response data should be serializable in json `JsonObjectType`. This is so that it can interoperate with other parts of the system such as the frontend.
+- You may conditionally check the `ServiceCall` property `return_response` to
+decide whether or not response data should be returned (e.g. if it is large or
+expensive to fetch). When not set, the service may return `None`.
+- Errors must be raised as exceptions, such as `HomeAssistantError`. The response data is not allowed to contain error codes or statuses which are used for error handling. We want to keep error handling simple to avoid error handling mistakes.
```

Comment:
```suggestion
- Errors must be raised as exceptions, such as `HomeAssistantError`. The response data is not allowed to contain error codes or statuses which are used for error handling.
```
Are other exception types allowed? If not, `errors must be raised as "homeassistanterrors"` would be much clearer.

I'm wondering if the second sentence means that a service call like `raw_command` (just a hypothetical one, allowing sending a raw request to the device) would not be allowed to return the device response to the user? If yes, that's unfortunate as I have seen such uses as a temporary fix until some features get implemented...

Reply:
I reworded this to try to make the intent more clear.

I don't want to specify all error handling best practices here, or establish new standards here, but just say we want the existing standards to apply here too. There is some discussion of this on https://developers.home-assistant.io/docs/integration_quality_scale_index but there are a couple cases that need further consistency / improvements.

For the raw command case: I don't have a strong opinion here, but it seems to be trying to solve something other than returning data for automations. Anyway, I've clarified the intent about not pushing complex error handling on to users as the main goal.